dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiPooling.scala

```diff
@@ -0,0 +1,218 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect._
+
+/**
+ * Region of interest pooling
+ * The RoIPooling uses max pooling to convert the features inside any valid region of interest
+ * into a small feature map with a fixed spatial extent of pooledH × pooledW (e.g., 7 × 7)
+ * an RoI is a rectangular window into a conv feature map.
+ * Each RoI is defined by a four-tuple (x1, y1, x2, y2) that specifies its
+ * top-left corner (x1, y1) and its bottom-right corner (x2, y2).
+ * RoI max pooling works by dividing the h × w RoI window into an pooledH × pooledW grid of
+ * sub-windows of approximate size h/H × w/W and then max-pooling the values in each sub-window
+ * into the corresponding output grid cell.
+ * Pooling is applied independently to each feature map channel
+ * @param pooledW
+ * @param pooledH
+ * @param spatialScale
+ * @tparam T Numeric type. Only support float/double now
+ */
+class RoiPooling[T: ClassTag](val pooledW: Int, val pooledH: Int, val spatialScale: T)
+  (implicit ev: TensorNumeric[T]) extends AbstractModule[Table, Tensor[T], T] {
+
+  private val argmax: Tensor[T] = Tensor[T]
+  private val gradInputTensor: Tensor[T] = Tensor[T]
+  gradInput.insert(gradInputTensor)
+
+  override def updateOutput(input: Table): Tensor[T] = {
+    require(input.length() == 2, "there must have two tensors in the table")
+
+    val data = input[Tensor[T]](1) // Input data to ROIPooling
+    val rois = input[Tensor[T]](2) // Input label to ROIPooling
+
+    require(rois.size().length > 1 && rois.size(2) == 5, "roi input shape should be (R, 5)")
+
+    output.resize(rois.size(1), data.size(2), pooledH, pooledW)
+      .fill(ev.fromType[Double](Double.MinValue))
+    val outputData = output.storage().array()
+    argmax.resizeAs(output).fill(ev.fromType(-1))
+    val argmaxData = argmax.storage().array()
+    val inputData = data.storage().array()
+    val dataOffset = offset(0, 1, sizes = data.size())
+    val argmaxOffset = offset(0, 1, sizes = argmax.size())
+    val outputOffset = offset(0, 1, sizes = output.size())
+    val roisOffset = offset(1, sizes = rois.size())
+    var n = 0
+    val dataSize = data.size()
+    while (n < rois.size(1)) {
+      poolOneRoi(n, rois(n + 1),
+        inputData, dataSize, dataOffset,
+        argmaxData, argmaxOffset,
+        outputData, outputOffset,
+        roisOffset)
+      n += 1
+    }
+    output
+  }
+
+  def poolOneRoi(n: Int, roi: Tensor[T],
```

Comment:
make this method private. and I suggest write two version of poolOneRoi: poolOneRoiDouble and poolOneRoiFloat. It's computation sensitive, don't use generic in it